task-140/server-socket

Misax148 opened the pull request:
# US Related

- [US-130](https://tree.taiga.io/project/joseluis-teran-coffeetime/us/130?milestone=394885)
- [Task-140](https://tree.taiga.io/project/joseluis-teran-coffeetime/task/140)

## Description

In this task, the server socket was implemented that will be waiting for the frames that the client socket will send from the console.

#### What did I do?
- I have created structures that represent the `Frame` that I will receive from the client.
- I am deserializing the json that the client sent me.
- I am converting the `Frame` to the necessary resources needed by the ui and sound in nannou.
- I have created the server socket through a thread within the nannou model so that it is constantly waiting for data and displays it in nannou. 
#### How did I do it?
- Use the `serde` library for deserialization.
- Use the methods already implemented by the UI and sound to represent it in nannou.
- Use the `threadpool` library to handle sound separately to avoid visual loss.
#### Why did I do it?
- To allow communication between the console and the screen.
### Notes
You need to add a file `.env` on the root of screen with the following information: 
`SCREEN_SOCKET_SERVER=/tmp/socket_console`

## Type of Change

Put an `x` in all the boxes that apply:

- [x] ✨ New Feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug Fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking Change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code Refactor
- [ ] ✅ Build Configuration Change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

## Pull Request Checklist

Put an `x` in all the boxes that apply:

- [x] My commit messages are detailed.
- [x] My code follows the code style of this project.
- [x] No existing features have been broken without good reason.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass.

zrock-dev reviewed (changes requested):
Following the [inet vs unix ](https://tree.taiga.io/project/joseluis-teran-coffeetime/task/141) conversation, please refactor this socket to use if_unix.

zrock-dev commented on src/socket_server/server.rs:
```diff
@@ -0,0 +1,32 @@
+    dotenv().ok();
+    let socket_path: String = env::var("ADDRESS").expect("ADDRESS must be set");
+
+    let _ = remove_file(&socket_path);
```
Why are you removing this file during [cold start](https://en.wikipedia.org/wiki/Cold_start_(computing))?

Misax148 replied:
In case there is any existing socket file in the same path to avoid link errors, just one extra step.

zrock-dev replied:
nice

zrock-dev commented on src/main.rs:
```diff
@@ -19,27 +39,54 @@ fn model(app: &App) -> Model {
+    let thread_pool = ThreadPoolBuilder::new().num_threads(4).build().unwrap();
 
-    Model { app_handler, sprite_list }  
+    Model {
```
is the model being used as a representation of the frame?

Misax148 replied:
No, the model in Nannou represents the management of the global state of the app, it manages the views, updates and events.

zrock-dev replied:
oke

zrock-dev commented on src/socket_server/server.rs:
```diff
@@ -0,0 +1,32 @@
+                    let reader = BufReader::new(stream);
+                    let frame: Frame = serde_json::from_reader(reader).unwrap();
+                    tx.send(frame).unwrap();
+                });
```
A simple game would send 60 frames per second, meaning spawning 60 frames per second

zrock-dev commented on src/socket_server/server.rs:
```diff
@@ -0,0 +1,32 @@
+                let tx = tx.clone();
+                thread::spawn(move || {
+                    let reader = BufReader::new(stream);
+                    let frame: Frame = serde_json::from_reader(reader).unwrap();
```
This logic is related to the receiver, can this be moved to another function?

zrock-dev reviewed (changes requested):
Spawning a thread per received frame would use a lot of resources.

zrock-dev commented:
Maybe it would be useful to use a LIFO queue to process those frames

![image](https://github.com/Pending-Name-21/screen/assets/64939916/02fccf42-4551-450c-84af-f41e1704a479)

Misax148 commented:
> Maybe it would be useful to use a LIFO queue to process those frames
> 
> ![image](https://private-user-images.githubusercontent.com/64939916/341096312-02fccf42-4551-450c-84af-f41e1704a479.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4MDU3MzUsIm5iZiI6MTcxODgwNTQzNSwicGF0aCI6Ii82NDkzOTkxNi8zNDEwOTYzMTItMDJmY2NmNDItNDU1MS00NTBjLTg0YWYtZjQxZTE3MDRhNDc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDEzNTcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3NTg0ZGIxYjEzYjAwYjYyZmRmNmRlYmY2OTdhNDdkODlhZDk4OTViYTJiZWMxNTdkY2JhMDQ1MTQ3NThhZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.75X-_JSBQMN9veI50pJryzb7tnNyM2U0R55VDX7cXhI)

I will try it

zrock-dev commented:
I was thinking that the logic of that component could be decoupled because it seems socket server is doing more than being a socket.
![image](https://github.com/Pending-Name-21/screen/assets/64939916/f436e071-5a27-4986-ba5d-f9618c0a6658)
![image](https://github.com/Pending-Name-21/screen/assets/64939916/15c874d3-7cae-4045-bc5d-b251f7062299)

Misax148 commented:
> I was thinking that the logic of that component could be decoupled because it seems socket server is doing more than being a socket. ![image](https://private-user-images.githubusercontent.com/64939916/341122896-f436e071-5a27-4986-ba5d-f9618c0a6658.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwMDc1NDQsIm5iZiI6MTcxOTAwNzI0NCwicGF0aCI6Ii82NDkzOTkxNi8zNDExMjI4OTYtZjQzNmUwNzEtNWEyNy00OTg2LWJhNWQtZjk2MThjMGE2NjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDIyMDA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzNzExNWY3NTM2ZWY1NzMyMTQ2OGQwOTY5NjNlMTgxNWUzOWNlZTIwMTJhNTg5OTNjZGZmNTIwNWFkODczZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-cExa0j1rFC7LdkaZJhBd5njqs19IUGdGgsdNn6ZaEQ) ![image](https://private-user-images.githubusercontent.com/64939916/341123028-15c874d3-7cae-4045-bc5d-b251f7062299.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwMDc1NDQsIm5iZiI6MTcxOTAwNzI0NCwicGF0aCI6Ii82NDkzOTkxNi8zNDExMjMwMjgtMTVjODc0ZDMtN2NhZS00MDQ1LWJjNWQtYjI1MWY3MDYyMjk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDIyMDA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkZTY4YWJmMmY3ZDA0M2Y5OGE5YWNlOTUzOWRmZTczY2JkMmNhYWI4NDNkY2NiN2ZjOTBjMDk0ZTYzY2NmZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sZWdoHxw9M5ZdF2OkCLVI2aWvqo47lNwcuz2TfynCPw)

I have implemented such as you recommended, but the only diference is that, i'm not implemented de LIFO algothim, instead of that I implemented FIFO, that is because with LIFO some Frames are lost during the deserialization and on the reprentation of Nannou is visible when this frame is lost

zrock-dev commented:
The client socket has been [implemented](https://github.com/Pending-Name-21/frontend-library/pull/25) using the [flatbuffesr](https://flatbuffers.dev/) library.

The IPC between screen and bridge is performance critical, and the current [serde implementation is slow](https://flatbuffers.dev/flatbuffers_benchmarks.html).
![image](https://github.com/Pending-Name-21/screen/assets/64939916/33f138b8-a338-4da7-b7cb-24a9748f1bf0)

In the other hand we need to have a communication protocol between these two containers, that is also [implemented](https://github.com/Pending-Name-21/miscellany/pull/15) using flatbuffers, which also has the generated files


I've made a spike regarding this implementation [spike-xx/serialization](https://github.com/Pending-Name-21/screen/tree/spike-xx/serialization) about how to pack into a flatbuffer.

I've used this [tutorial](https://flatbuffers.dev/flatbuffers_guide_tutorial.html)  to create the spike.

zrock-dev commented on src/socket_server/frame.rs:
```diff
@@ -0,0 +1,33 @@
+pub struct Frame {
+    pub sprite: Option<Sprite>,
+    pub sound: Option<Sound>,
+}
```
What about making sprite and sound non optional? or under which cases would it be None?

Misax148 replied:
I just changed to non optional both. Following the table defined for the `fbs` generated

zrock-dev commented on src/socket_server/server.rs:
```diff
@@ -0,0 +1,39 @@
+
+pub fn run_server(tx: Sender<Frame>, thread_pool: Arc<ThreadPool>) {
+    dotenv().ok();
+    let socket_path: String = env::var("ADDRESS").expect("ADDRESS must be set");
```
What about SCREEN_SOCKET_SERVER instead of ADDRESS so that the name is more descriptive

zrock-dev commented on src/socket_server/server.rs:
```diff
@@ -0,0 +1,39 @@
+                        if let Ok(frame) = serde_json::from_slice(&buffer) {
+                            let _ = tx.send(frame);
+                        }
+                    }
```
Good job with the performance optimization on concurrent socket client handling.

What about letting FrameReceiver deserialize the buffer?

zrock-dev reviewed (changes requested):
What if QueueManager uses Conversion only for conversion like FrameSprite to GuiSprite, and deliver those structs to their manager/handler like GuiHandler. This way, Conversion's responsibility would be to only convert structs.

![image](https://github.com/Pending-Name-21/screen/assets/64939916/6aabbac7-f887-4813-bd5f-c4f1386d5328)

Now, GUI Handler and Sound Manager don't have implementations to handle a struct, so they would have to be updated.

zrock-dev commented:
> > I was thinking that the logic of that component could be decoupled because it seems socket server is doing more than being a socket. ![image](https://private-user-images.githubusercontent.com/64939916/341122896-f436e071-5a27-4986-ba5d-f9618c0a6658.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwMDc1NDQsIm5iZiI6MTcxOTAwNzI0NCwicGF0aCI6Ii82NDkzOTkxNi8zNDExMjI4OTYtZjQzNmUwNzEtNWEyNy00OTg2LWJhNWQtZjk2MThjMGE2NjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDIyMDA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzNzExNWY3NTM2ZWY1NzMyMTQ2OGQwOTY5NjNlMTgxNWUzOWNlZTIwMTJhNTg5OTNjZGZmNTIwNWFkODczZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-cExa0j1rFC7LdkaZJhBd5njqs19IUGdGgsdNn6ZaEQ) ![image](https://private-user-images.githubusercontent.com/64939916/341123028-15c874d3-7cae-4045-bc5d-b251f7062299.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwMDc1NDQsIm5iZiI6MTcxOTAwNzI0NCwicGF0aCI6Ii82NDkzOTkxNi8zNDExMjMwMjgtMTVjODc0ZDMtN2NhZS00MDQ1LWJjNWQtYjI1MWY3MDYyMjk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDIyMDA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkZTY4YWJmMmY3ZDA0M2Y5OGE5YWNlOTUzOWRmZTczY2JkMmNhYWI4NDNkY2NiN2ZjOTBjMDk0ZTYzY2NmZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sZWdoHxw9M5ZdF2OkCLVI2aWvqo47lNwcuz2TfynCPw)
> 
> I have implemented such as you recommended, but the only diference is that, i'm not implemented de LIFO algothim, instead of that I implemented FIFO, that is because with LIFO some Frames are lost during the deserialization and on the reprentation of Nannou is visible when this frame is lost

Yep, you're right, the correct Data Structure would be FIFO

zrock-dev commented on src/socket_server/queue_manager.rs:
```diff
@@ -0,0 +1,36 @@
+                sprite_list.add_sprite(
+                    app,
+                    &frame_sprite.path,
+                    "received_sprite",
```
@codeFactory12 Why does the sprite_list is using a hashmap?

codeFactory12 replied:
for the update of the Sprites bassed on a same identification

zrock-dev commented on src/socket_server/deserialization.rs:
```diff
@@ -0,0 +1,36 @@
+    let sprites_vec: Vec<Sprite> = sprites.iter().map(|sprite| {
+        let position = sprite.position().ok_or_else(|| "Sprite position not found".to_string())?;
+        let size = sprite.size_().ok_or_else(|| "Sprite size not found".to_string())?;
+        Ok(Sprite {
+            position: Coord {
+                x: position.x(),
+                y: position.y(),
+            },
+            size: Size {
+                height: size.height(),
+                width: size.width(),
+            },
+            is_hidden: sprite.is_hidden(),
+            path: sprite.path().unwrap_or_default().to_string(),
```
What about using the structs from frame_generated.rs instead? I think this step can be avoided if so.

Misax148 replied:
Well, I tried it, and I've been having problems with the buffer not living long enough.
For the moment, the solution is to use our own structures that manage that lifetime, as long as there are no efficiency or performance problems during the execution of the game, this can be maintained for the moment. I will consult with the POO to keep an eye on it.

zrock-dev replied:
Yes, I agree. Maybe once a game is ready we will see it properly

zrock-dev reviewed (changes requested):
Good job. I just have a little observation.

zrock-dev commented:
I have created a UML based on the current relationship-
![image](https://github.com/Pending-Name-21/screen/assets/64939916/a6f52f8e-a86f-41da-b5ce-527490f4ff86)


I must say that I find this relationship odd, AudioManager is meant to handle audio and Socket Server should provide it with the data. What do you think @codeFactory12? 
![image](https://github.com/Pending-Name-21/screen/assets/64939916/bfa9de2c-026b-4866-bfdb-c703654b4fa9)

codeFactory12 commented:
> I have created a UML based on the current relationship- ![image](https://private-user-images.githubusercontent.com/64939916/343980816-a6f52f8e-a86f-41da-b5ce-527490f4ff86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NDYwNDcsIm5iZiI6MTcxOTU0NTc0NywicGF0aCI6Ii82NDkzOTkxNi8zNDM5ODA4MTYtYTZmNTJmOGUtYTg2Zi00MWRhLWI1Y2UtNTI3NDkwZjRmZjg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI4VDAzMzU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzYzA5ZDRjYzkwMTk3MjcyMjM1MTlkMGEyZWU0OGY2NDZiOWZiMTY1ODIzZDM4NWYwYjM1YjhhOWFjNGU1MzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.X7udglTyCL62AJ4OOnj4l-HKMOoFzi-DWmxE2cAh4u0)
> 
> I must say that I find this relationship odd, AudioManager is meant to handle audio and Socket Server should provide it with the data. What do you think @codeFactory12? ![image](https://private-user-images.githubusercontent.com/64939916/343980920-bfa9de2c-026b-4866-bfdb-c703654b4fa9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NDYwNDcsIm5iZiI6MTcxOTU0NTc0NywicGF0aCI6Ii82NDkzOTkxNi8zNDM5ODA5MjAtYmZhOWRlMmMtMDI2Yi00ODY2LWJmZGItYzcwMzY1NGI0ZmE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI4VDAzMzU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxMzZkNjlmZjMwODZkYTAzZWFiMTE2MmFiYmIyMTE0NTFlOGM2YmZhZDM1NDliMzA1ZDI1ZGIyYzcyZmZiNGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.goqg-TEOsoKuM75_z7AvycY1X50n6lQlV1uI4Jv5pIQ)

Yes it does but needs to be called the methods of  AudioManager manage all the logic to reproduce sounds bassed on a existent path, but needs to be called somewhere so can be able to be used

zrock-dev reviewed (approved):
Good job.
